elm/wizard.py

```diff
@@ -311,3 +238,310 @@ def chat(self, query,
+
+        Parameters
+        ----------
+        query : str
+            Question being asked of GPT
+        limit : int
+            Number of top results to return.
+
+        Returns
+        -------
+        strings : np.ndarray
+            1D array of related strings
+        score : np.ndarray
+            1D array of float scores of strings
+        idx : np.ndarray
+            1D array of indices in the text corpus corresponding to the
+            ranked strings/scores outputs.
+        """
+
+        embedding = self.get_embedding(query)
+        scores = 1 - self.cosine_dist(embedding)
+        best = np.argsort(scores)[::-1][:limit]
+
+        strings = self.text_arr[best]
+        scores = scores[best]
+
+        return strings, scores, best
+
+    def make_ref_list(self, idx):
+        """Make a reference list
+
+        Parameters
+        ----------
+        used_index : np.ndarray
+            Indices of the used text from the text corpus
+
+        Returns
+        -------
+        ref_list : list
+            A list of references (strs) used. This takes information straight
+            from ``ref_col``. Ideally, this is something like:
+            ["{ref_title} ({ref_url})"]
+        """
+        ref_list = ''
+        if self.ref_col is not None and self.ref_col in self.corpus:
+            ref_list = list(self.corpus[self.ref_col].iloc[idx].unique())
+
+        return ref_list
+
+
+class EnergyWizardPostgres(EnergyWizardBase):
+    """Interface to ask OpenAI LLMs about energy research.
+
+    This class is for execution with a postgres vector database.
+    Connecting to the database requires the use of the psycopg2
+    python package, environment variables storing the db user and
+    password, and the specification of other connection paremeters
+    such as host, port, and name. The database has the following
+    columns: id, embedding, chunks, and metadata.
+    """
+
+    def __init__(self, db_host, db_port, db_name,
+                 model=None, token_budget=3500):
+        """
+        Parameters
+        ----------
+        model : str
+            GPT model name, default is the DEFAULT_MODEL global var
+        token_budget : int
+            Number of tokens that can be embedded in the prompt. Note that the
+            default budget for GPT-3.5-Turbo is 4096, but you want to subtract
+            some tokens to account for the response budget.
+        db_host : str
+            Host url for postgres database.
+        db_port : str
+            Port for postres database. ex: '5432'
+        db_name : str
+            Postgres database name.
+        """
+        boto3 = try_import('boto3')
+        psycopg2 = try_import('psycopg2')
+
+        db_user = os.getenv("EWIZ_DB_USER")
+        db_password = os.getenv('EWIZ_DB_PASSWORD')
+        assert db_user is not None, "Must set user for postgres database!"
+        assert db_password is not None, "Must set user for postgres database!"
```

Comment:
These error messages should have the environment variable name so if not set the user knows what to set.